[homematic] Invalid configuration possible that disables updates from CCU3

## Expected Behavior
The configuration values for IP-Adresses and ports should be validated/automatically trimmed, so that an invalid configuration is not possible.

## Current Behavior
If the Callback-IP contains one or more blanks at the beginning or the end the thing goes online, but there occours an error on the CCU3 and no values are updated any more in openHAB.

## Possible Solution
- Validate, that IP and Port does not contains white-Spaces
- remove whitespaces from ip and configuration before usage.

## Steps to Reproduce (for Bugs)
1. Create CCU Thing
2. Configure Callback-URL with Whitespace
3. Thing States will not be updated

## Context
I've some homematic components, for example a sensor that shows, if the garage door is closed.
But the values for this won't be updated any more, when the configuration is invalid.

## Your Environment
`runtimeInfo:
  version: 4.1.3
  buildString: Release Build
locale: de-DE
systemInfo:
  configFolder: /etc/openhab
  userdataFolder: /var/lib/openhab
  logFolder: /var/log/openhab
  javaVersion: 17.0.5
  javaVendor: Eclipse Adoptium
  javaVendorVersion: Temurin-17.0.5+8
  osName: Linux
  osVersion: 6.1.39-v7l+
  osArchitecture: arm
  availableProcessors: 4
  freeMemory: 58138160
  totalMemory: 456130560
  uptime: 3261
  startLevel: 100
addons:
  - automation-jsscripting
  - binding-astro
  - binding-avmfritz
  - binding-denonmarantz
  - binding-deutschebahn
  - binding-ecovacs
  - binding-exec
  - binding-heos
  - binding-homematic
  - binding-icalendar
  - binding-modbus
  - binding-mqtt
  - binding-netatmo
  - binding-network
  - binding-ntp
  - binding-tankerkoenig
  - binding-tradfri
  - persistence-influxdb
  - persistence-rrd4j
  - transformation-jsonpath
  - transformation-map
  - transformation-regex
  - transformation-scale
  - ui-basic
clientInfo:
  device:
    ios: false
    android: false
    androidChrome: false
    desktop: true
    iphone: false
    ipod: false
    ipad: false
    edge: false
    ie: false
    firefox: false
    macos: false
    windows: true
    cordova: false
    phonegap: false
    electron: false
    nwjs: false
    webView: false
    webview: false
    standalone: false
    os: windows
    pixelRatio: 1.5
    prefersColorScheme: light
  isSecureContext: true
  locationbarVisible: true
  menubarVisible: true
  navigator:
    cookieEnabled: true
    deviceMemory: 8
    hardwareConcurrency: 8
    language: de-DE
    languages:
      - de-DE
      - de
      - en-US
      - en
    onLine: true
    platform: Win32
  screen:
    width: 2560
    height: 1440
    colorDepth: 24
  support:
    touch: false
    pointerEvents: true
    observer: true
    passiveListener: true
    gestures: false
    intersectionObserver: true
  themeOptions:
    dark: light
    filled: true
    pageTransitionAnimation: default
    bars: filled
    homeNavbar: default
    homeBackground: default
    expandableCardAnimation: default
  userAgent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML,
    like Gecko) Chrome/126.0.0.0 Safari/537.36
timestamp: 2024-06-26T12:03:40.950Z
`

Stack Trace (in homematic) looks like this
```
2024-06-26 13:39:03,016 io.vertx.core.impl.ContextImpl ERROR [vert.x-worker-thread-19] Unhandled exception
java.lang.IllegalArgumentException: Illegal character in authority at index 7: http:// 192.168.XXX.XXX:9125
        at java.net.URI.create(Unknown Source) ~[?:?]
        at org.apache.http.client.methods.HttpPost.<init>(HttpPost.java:76) ~[HMIPServer.jar:?]
        at de.eq3.cbcs.legacy.communication.rpc.internal.transport.http.HttpTransport.sendRequest(HttpTransport.java:104) ~[HMIPServer.jar:?]
        at de.eq3.cbcs.legacy.communication.rpc.internal.rpc.RpcClient.sendRequest(RpcClient.java:94) ~[HMIPServer.jar:?]
        at de.eq3.cbcs.legacy.communication.rpc.internal.rpc.RpcClient.invoke(RpcClient.java:82) ~[HMIPServer.jar:?]
        at com.sun.proxy.$Proxy39.listDevices(Unknown Source) ~[?:?]
        at de.eq3.ccu.virtualdevice.service.internal.rega.BackendUpdateDevicesCommand.updateDevicesForClient(BackendUpdateDevicesCommand.java:34) ~[HMIPServer.jar:?]
        at de.eq3.ccu.virtualdevice.service.internal.rega.BackendUpdateDevicesCommand.run(BackendUpdateDevicesCommand.java:29) ~[HMIPServer.jar:?]
        at de.eq3.ccu.virtualdevice.service.internal.rega.BackendWorker.handle(BackendWorker.java:27) ~[HMIPServer.jar:?]
        at de.eq3.ccu.virtualdevice.service.internal.rega.BackendWorker.handle(BackendWorker.java:11) ~[HMIPServer.jar:?]
        at io.vertx.core.impl.AbstractContext.dispatch(AbstractContext.java:100) ~[HMIPServer.jar:?]
        at io.vertx.core.impl.WorkerContext.lambda$emit$0(WorkerContext.java:59) ~[HMIPServer.jar:?]
        at io.vertx.core.impl.WorkerContext.lambda$execute$2(WorkerContext.java:104) ~[HMIPServer.jar:?]
        at io.vertx.core.impl.TaskQueue.run(TaskQueue.java:76) ~[HMIPServer.jar:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) ~[?:?]
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) ~[HMIPServer.jar:?]
        at java.lang.Thread.run(Unknown Source) ~[?:?]
Caused by: java.net.URISyntaxException: Illegal character in authority at index 7: http:// 192.168.XXX.XXX:9125
        at java.net.URI$Parser.fail(Unknown Source) ~[?:?]
        at java.net.URI$Parser.parseAuthority(Unknown Source) ~[?:?]
        at java.net.URI$Parser.parseHierarchical(Unknown Source) ~[?:?]
        at java.net.URI$Parser.parse(Unknown Source) ~[?:?]
        at java.net.URI.<init>(Unknown Source) ~[?:?]
        ... 18 more

```

--- a/bundles/org.openhab.binding.homematic/README.md
+++ b/bundles/org.openhab.binding.homematic/README.md
@@ -141,7 +141,7 @@ Network address of the Homematic gateway
 Hint for the binding to identify the gateway type (auto|ccu|noccu) (default = "auto").
 
 - **callbackHost**
-Callback network address of the system runtime, default is auto-discovery
+Callback network address of the system runtime, default is auto-discovery. This value must not contain any white spaces.
 
 - **xmlCallbackPort**
 Callback port of the binding's XML-RPC server, default is 9125 and counts up for each additional bridge

--- a/bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/handler/HomematicBridgeHandler.java
+++ b/bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/handler/HomematicBridgeHandler.java
@@ -20,6 +20,7 @@ import java.util.Collection;
 import java.util.Map;
 import java.util.Set;
 import java.util.concurrent.Future;
+import java.util.concurrent.ScheduledExecutorService;
 import java.util.concurrent.TimeUnit;
 
 import org.eclipse.jdt.annotation.NonNull;
@@ -58,11 +59,14 @@ import org.slf4j.LoggerFactory;
  * @author Gerhard Riegler - Initial contribution
  */
 public class HomematicBridgeHandler extends BaseBridgeHandler implements HomematicGatewayAdapter {
+
+    protected ScheduledExecutorService executorService = scheduler;
+
     private final Logger logger = LoggerFactory.getLogger(HomematicBridgeHandler.class);
     private static final long REINITIALIZE_DELAY_SECONDS = 10;
     private static final int DUTY_CYCLE_RATIO_LIMIT = 99;
     private static final int DUTY_CYCLE_DISCONNECTED = -1;
-    private static SimplePortPool portPool = new SimplePortPool();
+    private static final SimplePortPool portPool = new SimplePortPool();
 
     private final Object dutyCycleRatioUpdateLock = new Object();
     private final Object initDisposeLock = new Object();
@@ -93,7 +97,7 @@ public class HomematicBridgeHandler extends BaseBridgeHandler implements Homemat
     public void initialize() {
         synchronized (initDisposeLock) {
             isDisposed = false;
-            initializeFuture = scheduler.submit(this::initializeInternal);
+            initializeFuture = executorService.submit(this::initializeInternal);
         }
     }
 
@@ -106,6 +110,8 @@ public class HomematicBridgeHandler extends BaseBridgeHandler implements Homemat
             config = createHomematicConfig();
 
             try {
+                this.checkForConfigurationErrors();
+
                 String id = getThing().getUID().getId();
                 gateway = HomematicGatewayFactory.createGateway(id, config, this, httpClient);
                 configureThingProperties();
@@ -140,6 +146,15 @@ public class HomematicBridgeHandler extends BaseBridgeHandler implements Homemat
         }
     }
 
+    /**
+     * Validates, if the configuration contains errors.
+     */
+    private void checkForConfigurationErrors() {
+        if (this.config.getCallbackHost().contains(" ")) {
+            throw new ConfigurationException("The callback host mut not contain white spaces.");
+        }
+    }
+
     private void configureThingProperties() {
         final HmGatewayInfo info = config.getGatewayInfo();
         final Map<String, String> properties = getThing().getProperties();
@@ -160,7 +175,7 @@ public class HomematicBridgeHandler extends BaseBridgeHandler implements Homemat
      */
     private void scheduleReinitialize() {
         if (!isDisposed) {
-            initializeFuture = scheduler.schedule(this::initializeInternal, REINITIALIZE_DELAY_SECONDS,
+            initializeFuture = executorService.schedule(this::initializeInternal, REINITIALIZE_DELAY_SECONDS,
                     TimeUnit.SECONDS);
         }
     }
@@ -434,7 +449,7 @@ public class HomematicBridgeHandler extends BaseBridgeHandler implements Homemat
      * @param defer <i>true</i> will delete the device once it becomes available.
      */
     public void deleteFromGateway(String address, boolean reset, boolean force, boolean defer) {
-        scheduler.submit(() -> {
+        executorService.submit(() -> {
             logger.debug("Deleting the device '{}' from gateway '{}'", address, getBridge());
             getGateway().deleteDevice(address, reset, force, defer);
         });

--- a/bundles/org.openhab.binding.homematic/src/test/java/org/openhab/binding/homematic/internal/handler/HomematicBridgeHandlerMock.java
+++ b/bundles/org.openhab.binding.homematic/src/test/java/org/openhab/binding/homematic/internal/handler/HomematicBridgeHandlerMock.java
@@ -1,0 +1,45 @@
+/**
+ * Copyright (c) 2010-2024 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.homematic.internal.handler;
+
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.doAnswer;
+
+import java.util.concurrent.ScheduledExecutorService;
+
+import org.eclipse.jdt.annotation.NonNull;
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jetty.client.HttpClient;
+import org.mockito.Mockito;
+import org.mockito.invocation.InvocationOnMock;
+import org.openhab.binding.homematic.internal.type.HomematicTypeGenerator;
+import org.openhab.core.thing.Bridge;
+
+/**
+ * The {@link HomematicBridgeHandlerMock} is responsible for mocking {@link HomematicBridgeHandler}
+ *
+ * @author Sönke Küper - Initial contribution
+ */
+@NonNullByDefault
+public class HomematicBridgeHandlerMock extends HomematicBridgeHandler {
+
+    public HomematicBridgeHandlerMock(@NonNull Bridge bridge, HomematicTypeGenerator typeGenerator, String ipv4Address,
+            HttpClient httpClient) {
+        super(bridge, typeGenerator, ipv4Address, httpClient);
+        executorService = Mockito.mock(ScheduledExecutorService.class);
+        doAnswer((InvocationOnMock invocation) -> {
+            ((Runnable) invocation.getArguments()[0]).run();
+            return null;
+        }).when(executorService).submit(any(Runnable.class));
+    }
+}

--- a/bundles/org.openhab.binding.homematic/src/test/java/org/openhab/binding/homematic/internal/handler/HomematicBridgeHandlerTest.java
+++ b/bundles/org.openhab.binding.homematic/src/test/java/org/openhab/binding/homematic/internal/handler/HomematicBridgeHandlerTest.java
@@ -1,0 +1,65 @@
+/**
+ * Copyright (c) 2010-2024 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.homematic.internal.handler;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.mockito.ArgumentMatchers.argThat;
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
+
+import java.util.Objects;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jetty.client.HttpClient;
+import org.junit.jupiter.api.Test;
+import org.mockito.Mockito;
+import org.openhab.binding.homematic.internal.HomematicBindingConstants;
+import org.openhab.binding.homematic.internal.type.HomematicTypeGenerator;
+import org.openhab.core.thing.Bridge;
+import org.openhab.core.thing.ThingStatus;
+import org.openhab.core.thing.ThingStatusDetail;
+import org.openhab.core.thing.binding.ThingHandlerCallback;
+import org.openhab.core.thing.internal.BridgeImpl;
+
+/**
+ * @author Sönke Küper - Initial contribution
+ */
+@NonNullByDefault
+public class HomematicBridgeHandlerTest {
+
+    @Test
+    public void testGetRpcCallbackUrlDoesNotContainsSpaces() {
+        HttpClient httpClient = Mockito.mock(HttpClient.class);
+
+        Bridge bridge = new BridgeImpl(HomematicBindingConstants.THING_TYPE_BRIDGE, "1234");
+        bridge.getConfiguration().put("callbackHost", " 192. 168.1.1");
+        assertThat(bridge.getStatus(), is(ThingStatus.UNINITIALIZED));
+        HomematicTypeGenerator typeGenerator = mock(HomematicTypeGenerator.class);
+
+        HomematicBridgeHandlerMock handler = new HomematicBridgeHandlerMock(bridge, typeGenerator, "1.2.3.4",
+                httpClient);
+        ThingHandlerCallback callback = mock(ThingHandlerCallback.class);
+        handler.setCallback(callback);
+        handler.initialize();
+
+        try {
+            verify(callback).statusUpdated(eq(bridge), argThat(arg -> arg.getStatus().equals(ThingStatus.OFFLINE)
+                    && arg.getStatusDetail().equals(ThingStatusDetail.CONFIGURATION_ERROR)
+                    && Objects.equals(arg.getDescription(), "The callback host mut not contain white spaces.")));
+        } finally {
+            handler.dispose();
+        }
+    }
+}

